ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/TransactionSimulator.java

```diff
@@ -144,21 +144,31 @@ public Optional<TransactionSimulatorResult> process(
       return Optional.empty();
     }
 
-    WorldUpdater updater;
-    try {
-      updater = getWorldUpdater(header);
-    } catch (final IllegalArgumentException e) {
-      return Optional.empty();
-    }
+    try (var ws = getWorldState(header)) {
+
+      WorldUpdater updater = getEffectiveWorldStateUpdater(header, ws);
+
+      // in order to trace the state diff we need to make sure that
+      // the world updater always has a parent
+      if (operationTracer instanceof DebugOperationTracer) {
+        updater = updater.parentUpdater().isPresent() ? updater : updater.updater();
+      }
+
+      return processWithWorldUpdater(
+          callParams, transactionValidationParams, operationTracer, header, updater);
 
-    // in order to trace the state diff we need to make sure that
-    // the world updater always has a parent
-    if (operationTracer instanceof DebugOperationTracer) {
-      updater = updater.parentUpdater().isPresent() ? updater : updater.updater();
+    } catch (final Exception e) {
+      return Optional.empty();
     }
+  }
 
-    return processWithWorldUpdater(
-        callParams, transactionValidationParams, operationTracer, header, updater);
+  public MutableWorldState getWorldState(final BlockHeader header) {
+    return worldStateArchive
+        .getMutable(header.getStateRoot(), header.getHash(), false)
+        .orElseThrow(
+            () ->
+                new IllegalArgumentException(
+                    "Public world state not available for block " + header.getNumber()));
```

Comment:
this should be private and the impl of traceCallMany should get pushed into this class